pkg/doublewrite/doublewritebackend.go

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package doublewrite
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/allocator"
+	"github.com/cilium/cilium/pkg/idpool"
+	"github.com/cilium/cilium/pkg/k8s/identitybackend"
+	"github.com/cilium/cilium/pkg/kvstore"
+	kvstoreallocator "github.com/cilium/cilium/pkg/kvstore/allocator"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/rate"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "double-write-allocator")
+)
+
+func NewDoubleWriteBackend(c DoubleWriteBackendConfiguration) (allocator.Backend, error) {
+	crdBackend, _ := identitybackend.NewCRDBackend(c.CRDBackendConfiguration)
+	kvstoreBackend, _ := kvstoreallocator.NewKVStoreBackend(c.KVStoreBackendConfiguration)
+
+	log.WithFields(logrus.Fields{
+		"KVStoreBackendConfiguration.Suffix":   c.KVStoreBackendConfiguration.Suffix,
+		"KVStoreBackendConfiguration.Typ":      c.KVStoreBackendConfiguration.Typ.String(),
+		"KVStoreBackendConfiguration.BasePath": c.KVStoreBackendConfiguration.BasePath,
+		"readFromKVStore":                      c.ReadFromKVStore,
+	}).Debug("Creating the Double-Write backend")
+
+	return &doubleWriteBackend{crdBackend: crdBackend.(*identitybackend.CRDBackend), kvstoreBackend: kvstoreBackend, readFromKVStore: c.ReadFromKVStore}, nil
+}
+
+type DoubleWriteBackendConfiguration struct {
+	CRDBackendConfiguration     identitybackend.CRDBackendConfiguration
+	KVStoreBackendConfiguration kvstoreallocator.KVStoreBackendConfiguration
+	ReadFromKVStore             bool
+}
+
+type doubleWriteBackend struct {
+	crdBackend      *identitybackend.CRDBackend
+	kvstoreBackend  *kvstoreallocator.KVStoreBackend
+	readFromKVStore bool
+}
+
+func (d *doubleWriteBackend) DeleteAllKeys(ctx context.Context) {
+	d.crdBackend.DeleteAllKeys(ctx)
+	d.kvstoreBackend.DeleteAllKeys(ctx)
+}
+
+func (d *doubleWriteBackend) AllocateID(ctx context.Context, id idpool.ID, key allocator.AllocatorKey) (allocator.AllocatorKey, error) {
+	crdKey, crdErr := d.crdBackend.AllocateID(ctx, id, key)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr).Error("CRD backend failed to allocate identity")
+	}
+	kvStoreKey, kvStoreErr := d.kvstoreBackend.AllocateID(ctx, id, key)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to allocate identity")
+	}
+	if d.readFromKVStore {
+		return kvStoreKey, kvStoreErr
+	}
+	return crdKey, crdErr
```

Comment:
So I think trying to make these operations atomic would make the code exponentially more complex but wouldn't add a whole lot of benefit. 

You would need a system to rollback to a previous state if one of the writes failed: for object creations that would probably be possible with careful deletes but for object updates you would need to store the previous state in memory and also account for race conditions due to concurrent updates... IMO that's too complex, especially for a backend which is supposed to be used only for a short period of time.

Moreover, the current code already handles failed writes to the "primary" backend (i.e. the one used for reads) and there is alerting which helps with detecting discrepancies (in case a write to the "primary" backend succeeded but failed for the "secondary" backend).

Reply:
> So I think trying to make these operations atomic would make the code exponentially more complex but wouldn't add a whole lot of benefit.

Yeah, fair enough. I agree that it makes sense to skip that considering that it may not be possible to actually revert the previous changes, or lead to additional inconsistencies.

I still wonder whether it would be better to return an error if any of the two backend operations failed though, as that would allow the caller to detect that something went wrong, and perform a retry. My assumption is that it should help to automatically resolve temporary errors (e.g., kvstore blips, or creation conflicts), which may otherwise be retried only during a subsequent agent restart. WDYT?

Reply:
I looked into it a bit more and I could technically refactor the code to return errors with something like this:
```diff
@@ -122,10 +126,12 @@ func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key al
        crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
        if crdErr != nil {
                log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
+               return crdErr
        }
        kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
        if kvStoreErr != nil {
                log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
+               return kvStoreErr
        }
```
However, this adds a small complication: now if a CRD update fails, the kvstore update will never go through. My initial idea was that in `doublewrite-readkvstore` mode, the kvstore writes should be the "primary" ones and CRD writes are made on a best-effort basis, trying to not interfere with kvstore operations (and vice-versa in `doublewrite-readcrd` mode). This way the `doublewrite-readkvstore` functions in the same way as the regular `kvstore` mode but with best-effort CRD writes in parallel.

I could refactor the code even more to take this into account:
```go
func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
	if d.readFromKVStore {
		// If we are reading from the kvstore, we perform the kvstore update first
		kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
		if kvStoreErr != nil {
			log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
			return kvStoreErr
		}
		crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
		if crdErr != nil {
			log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
			return crdErr
		}
	}
	// If we are reading from CRDs, we perform the CRD update first
	crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
	if crdErr != nil {
		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
		return crdErr
	}
	kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
	if kvStoreErr != nil {
		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
		return kvStoreErr
	}
	return nil
}
```
But this makes the code more convoluted (I'll need to write the same logic for all write methods basically). What's your take on this?

Reply:
Because migration is expected to be reversible (as outlined in the original proposal), we must always write to both datastores, as @giorio94 observes.

This is especially critical given that different agents may have a different understanding of "primary". As configuration changes are rolled out, *if* we were to ignore secondary-side failures, we could have pretty serious divergence.

Given that most migrations are probably to the CRD store, and if the k8s apiserver goes down, nothing is happening anyways, I believe the right pattern is to write to CRD first, then kvstore. But, if kvstore fails, we must revert our CRD change and return failure.

There are still some nasty corner cases around interleaving failures between doublewrite and non-doublewrite agents, but I think those are very unlikely.

Reply:
Makes sense, I wrote some code to handle this for the allocation functions: https://github.com/cilium/cilium/pull/31920/files#diff-ce8e3e1201058922bca6b6b8fd0ffc3a9ac595a34621246e05daf000f343bcb5R80-R100 
LMK what you think.

And also added a test: https://github.com/cilium/cilium/pull/31920/files#diff-08f67e47fb93b88a978cbf9b31e32f4af4b01ac36b1a78b9890201cf654b8d6aR89-R109

I didn't add the same logic for the `UpdateKey` / `AcquireReference` functions because rolling back the CRD update is not that important there: these functions basically update the heartbeat timestamp on the CRD object, so in the worst case, the GC of the identity will be delayed.